Source/SuperLinq.Async/SortedMerge.cs

```diff
@@ -291,39 +288,91 @@ public static partial class AsyncSuperEnumerable
+			var enumerators = new List<IAsyncEnumerator<TSource>>();
 
-			// prime all of the iterators by advancing them to their first element (if any)
-			for (var i = 0; await list.MoveNext(i).ConfigureAwait(false); i++)
-			{ }
-
-			// while all iterators have not yet been consumed...
-			while (list.Any())
+			try
 			{
-				var nextIndex = 0;
-				var nextValue = list.Current(0);
-				var nextKey = keySelector(nextValue);
+				// Ensure we dispose first N enumerators if N+1 throws 
+				foreach (var sequence in sequences)
+				{
+					var e = sequence.GetAsyncEnumerator(cancellationToken);
+					if (await e.MoveNextAsync())
+						enumerators.Add(e);
+					else
+						await e.DisposeAsync();
+				}
+#if NET6_0_OR_GREATER
+				var queue = new PriorityQueue<IAsyncEnumerator<TSource>, TKey>(
+					enumerators.Select(x => (x, keySelector(x.Current))),
+					comparer);
+
+#pragma warning disable CA2000 // e will be disposed via enumerators list
+				while (queue.TryDequeue(out var e, out var _))
+#pragma warning restore CA2000 // Dispose objects before losing scope
+				{
+					yield return e.Current;
+
+					// Fast drain of final enumerator
+					if (queue.Count == 0)
+					{
+						while (await e.MoveNextAsync()) yield return e.Current;
+						break;
+					}
+
+					if (await e.MoveNextAsync()) queue.Enqueue(e, keySelector(e.Current));
+				}
+
+#else
+				enumerators.Sort((x, y) => comparer.Compare(keySelector(x.Current), keySelector(y.Current)));
 
-				// find the next least element to return
-				for (var i = 1; i < list.Count; i++)
+				var arr = enumerators.ToArray();
+				var count = arr.Length;
+				var sourceComparer = new SourceComparer<TSource, TKey>(comparer, keySelector);
+
+				while (count > 1)
 				{
-					var anotherElement = list.Current(i);
-					var anotherKey = keySelector(anotherElement);
-					// determine which element follows based on ordering function
-					if (comparer.Compare(nextKey, anotherKey) > 0)
+					var e = arr[0];
+					yield return e.Current;
+
+					if (!await e.MoveNextAsync())
 					{
-						nextIndex = i;
-						nextValue = anotherElement;
-						nextKey = anotherKey;
+						count--;
+						Array.Copy(arr, 1, arr, 0, count);
+						continue;
 					}
+
+					var index = Array.BinarySearch(arr, 1, count - 1, e, sourceComparer);
+					if (index < 0) index = ~index;
+
+					index--;
+					if (index > 0) Array.Copy(arr, 1, arr, 0, index);
+					arr[index] = e;
 				}
 
-				yield return nextValue; // next value in precedence order
+				if (count == 1)
+				{
+					var e = arr[0];
+					yield return e.Current;
 
-				// advance iterator that yielded element, excluding it when consumed
-				_ = await list.MoveNextOnce(nextIndex).ConfigureAwait(false);
+					while (await e.MoveNextAsync()) yield return e.Current;
+				}
+#endif
+			}
+			finally
+			{
+				foreach (var e in enumerators) await e.DisposeAsync();
 			}
 		}
 	}
+
+#if !NET6_0_OR_GREATER
+	internal sealed record class SourceComparer<TItem, TKey>(
```

Comment:
The style guideline failed because this is still a `record class`, when it should be a `class`.